src/routes/files.ts

```diff
@@ -1,14 +1,412 @@
+  timestamp: number
+  paths: {
+    [path: string]: number
+  }
+}
+
+const VALID_PATH_REGEX = /^(\/([^\/ ]+([ ]+[^\/ ]+)*)+)+\/?$/
 
 export const filesRouter = Router()
 
+// ---------------------------------------------------------------------
+// POST /files
+// ---------------------------------------------------------------------
+
+filesRouter.post('/files', async (req, res, next) => {
+  let body: FilesPostBody
+  let pathKeys: string[]
+  let repoKey: string
+
+  // Validate request body
+  try {
+    body = asFilesPostBody(req.body)
+
+    pathKeys = Object.keys(body.paths)
+
+    const repoid = pathKeys[0].split(':')[0]
+    repoKey = `${repoid}:/`
+
+    // Validate fileKeys are formated correctly; includes repoid and valid
+    // paths (no double slashes etc).
+    pathKeys.forEach(pathKey => {
+      const [repoid, path] = pathKey.split(':')
+
+      if (repoKey !== `${repoid}:/`) {
+        const apiClientError: ApiClientError = {
+          status: 400,
+          message: `All paths must be of the same repo.`
+        }
+        throw apiClientError
+      }
+
+      if (!repoid) {
+        const apiClientError: ApiClientError = {
+          status: 400,
+          message: `Missing repoid for key: '${pathKey}'`
+        }
+        throw apiClientError
+      }
+      if (!path) {
+        const apiClientError: ApiClientError = {
+          status: 400,
+          message: `Missing path for key: '${pathKey}'`
+        }
+        throw apiClientError
+      }
+      if (path === '/') {
+        const apiClientError: ApiClientError = {
+          status: 400,
+          message:
+            `Invalid path '${path}' for key '${pathKey}'. ` +
+            `Path cannot be root.`
+        }
+        throw apiClientError
+      }
+      if (!VALID_PATH_REGEX.test(path)) {
+        const apiClientError: ApiClientError = {
+          status: 400,
+          message: `Invalid path '${path}' for key '${pathKey}'. `
+        }
+        throw apiClientError
+      }
+    })
+  } catch (error) {
+    const response: ApiErrorResponse = {
+      success: false,
+      message: error.message
+    }
+    return res.status(400).json(response)
+  }
+
+  // Validate request body timestamp
+  let repoDoc: StoreRootDocument
+  try {
+    const repoQuery = await dataStore.get(repoKey)
+    repoDoc = asStoreRootDocument(repoQuery)
+  } catch (err) {
+    throw new Error(`Failed to validate repo timestamp: ${err.message}`)
+  }
+
+  if (repoDoc.timestamp !== body.timestamp) {
+    const apiClientError: ApiClientError = {
+      status: 422,
+      message: `Failed due to out-of-date timestamp`
+    }
+    throw apiClientError
+  }
+
+  // Timestamp is the same for all updates for this request
+  let requestTimestamp = Date.now()
+
+  while (true) {
+    try {
+      const success = await filesUpdateRoutine(body, requestTimestamp)
+
+      if (success) {
+        break
+      } else {
+        requestTimestamp = Date.now()
+      }
```

Comment:
Remove `success` boolean in favor of a custom error thrown for rev-mismatch errors.

Reply:
Add a max retry of 100 times. If retries fail, throw server error.

Reply:
Resolution: https://github.com/EdgeApp/edge-sync-server/commit/e41d00252299e57f88a240b02f1eb37dbaaf2842#diff-2196a720a13a3fd21839e328edb702c97e36909db9cc12f2175981232fdc3ce7R102